Python/segment_tree.py

```diff
@@ -0,0 +1,156 @@
+import math
+
+
+class SegmentTree:
+    '''Class representation of Segment Tree
+    
+    ...
+
+    Attributes:
+    ----------
+    arr: Any iterable
+        the segment tree will be built using this iterable
+    len-arr: int
+        length of the iterable arr
+    func: method/function
+        the function that will be used to built the tree
+    default: Any
+        the default value with which the tree array is initialized
+    tree: list
+        the representation of the tree
+
+    Note: Indexing is 0 based in this implementation
+    '''
+    def __init__(self, arr, func, default):
+
+        # Checking if iterable passed is of NoneType
+        if arr is not None:
+            self.arr = arr
+            self.len_arr = len(self.arr)
+        else:
+            raise TypeError('Argument arr should not be null')
+
+        # Checking if function passed is of NoneType
+        if func is not None:
+            self.func = func
+        else:
+            raise TypeError('Argument func should not be null')
+
+        self.default = default
+        self._build_tree()
+
+    def _is_power_of_2(self, n):
+        '''Function to check if n is already a power of 2.
+
+        Equation: 2ˣ = n
+        After log transformation: x = log₂(n)/log₂(2)
+        '''
+        def log_base_2(n):
+            return math.log10(n) / math.log10(2) if n != 0 else 0
+
+        return math.ceil(log_base_2(n)) == math.floor(log_base_2(n))
+
+    def _next_power_of_2(self, n):
+        '''Function to return the next number which is a power of 2'''
+
+        # Checks if n is already a power of 2
+        if self._is_power_of_2(n):
+            return n
+
+        val = 1
+
+        # Loop will run until it give us the value which is greater than n.
+        while val <= n:
+
+            # val will be left shifted by 1 in each iteration.
+            val = val << 1
+        return val
+
+    def _build_tree(self):
+        '''Function to build the segment tree using the function provided.'''
+
+        # If number of leaves in a full binary tree is n,
+        # total number of internal nodes in the tree is n - 1
+        # Hence, total number of nodes in the tree will be 2n - 1.
+        self.n_leaves = self._next_power_of_2(self.len_arr)
+        n = self.n_leaves << 1
+
+        # initializing all nodes of the tree to 0
+        self.tree = [self.default for i in range(n)]
```

Comment:
`[self.default] * n` is a faster alternative. Also, loop variable i is unused. Use _ instead of i if you want to keep the same style of initializing a list.